examples/computer_vision/vision_transformer/det_callback.py

```diff
@@ -0,0 +1,150 @@
+import typing
+
+from transformers import (
+    TrainerCallback,
+    TrainerState,
+    TrainerControl,
+    TrainingArguments,
+)
+
+import os
+import determined as det
+import logging
+
+import weakref
+import importlib.util
+
+
+class DetCallback(TrainerCallback):
+
+    def __init__(self, args: typing.Dict, filter_metrics: typing.List = None,
+                 tokenizer: typing.Any = None, tokenizer_options: typing.Dict = None,
+                 checkpoint_metadata: typing.Dict = None) -> None:
+        super().__init__()
+
+        assert is_determined_available(), "DetCallback requires determined to be installed. Run `pip install determined`."
+        import determined
+        self._det = determined
+        distributed = self._det.core.DistributedContext.from_torch_distributed()
+        self.core_context = self._det.core.init(distributed=distributed)
+        self.core_context.__enter__()
+        weakref.finalize(self, exit_context, self.core_context)
+
+        self.filter_metrics = filter_metrics
+        self.user_checkpoint_metadata = checkpoint_metadata
+        self.tokenizer = tokenizer
+        self.tokenizer_options = tokenizer_options
+        self.load_last_checkpoint(args)
+
+        self.last_eval_metrics = None
+        self.searcher_metric = self._det.get_cluster_info().trial._config['searcher']['metric']
+        self.searcher_ops = self.core_context.searcher.operations()
+        self.current_op = next(self.searcher_ops)
+
+    def on_log(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, model=None, logs=None,
+               **kwargs):
+        if state.is_world_process_zero:
+            metrics = self.process_log(logs)
+            if 'loss' in metrics:
```

Comment:
Is `loss` or `eval_loss` always present?  Might be more resilient to check some other way whether we're in train or val -- looks like maybe `state.epoch is None` when we're in val?

Reply:
Good point.

[`Loss` is always present in the log](https://github.com/huggingface/transformers/blob/a123eee9df11b5a116452c3f539f2ff0427324d2/src/transformers/trainer.py#L2038) and refers to the loss during training. After HF Trainer finishes training, it also reports `train_loss` (average loss), which may cause DB error, if we try to report `train_loss` with the same step as previously reported `loss`. I decided to omit `train_loss`, but I am open to suggestions. Be aware that it's a single point and looks suspicious in our WebUI.

`eval_loss` is more tricky. For the evaluation step, we would get `eval_loss` [assuming loss is not None.](https://github.com/huggingface/transformers/blob/a123eee9df11b5a116452c3f539f2ff0427324d2/src/transformers/trainer.py#L3059). 
I see that WandB callback and TensorBoard callback use [re-write function](https://github.com/huggingface/transformers/blob/a123eee9df11b5a116452c3f539f2ff0427324d2/src/transformers/integrations.py#L461), that checks for prefixes, such as `eval_` or `test_`. I will follow the same approach and check for prefix since it seems safer than an exact check for `eval_loss`.

Reply:
`state.epoch` is not None when we run validation during training. I decided to do sth similar to the [re-write function](https://github.com/huggingface/transformers/blob/a123eee9df11b5a116452c3f539f2ff0427324d2/src/transformers/integrations.py#L461).